dowhy/gcm/model_evaluation.py

```diff
@@ -0,0 +1,794 @@
+
+    if evaluate_overall_kl_divergence:
+        # Normally, we need to estimate the KL divergence jointly. However, to avoid issues with high dimensional data,
+        # we approximate it by taking the average over the marginal KL divergences.
+        drawn_samples = draw_samples(causal_model, data.shape[0])
+
+        evaluation_result.overall_kl_divergence = np.mean(
+            [
+                auto_estimate_kl_divergence(drawn_samples[node].to_numpy(), data[node].to_numpy())
+                for node in causal_model.graph.nodes
+            ]
+        )
+
+    if evaluate_causal_structure:
+        evaluation_result.graph_falsification = falsify_graph(
+            causal_model.graph,
+            data,
+            n_permutations=config.max_num_permutations_falsify,
+            independence_test=config.independence_test_falsify,
+            conditional_independence_test=config.conditional_independence_test_falsify,
+            n_jobs=config.n_jobs,
+        )
+
+    return evaluation_result
+
+
+def _evaluate_model_performances(
+    causal_model: ProbabilisticCausalModel,
+    data: pd.DataFrame,
+    compare_mechanism_baselines,
+    baseline_models_regression: List[Callable[[], PredictionModel]],
+    baseline_models_classification: List[Callable[[], PredictionModel]],
+    kfolds: int,
+    n_jobs: int,
+) -> Dict[Any, Tuple[bool, float, Optional[int], Optional[str], int, Optional[float]]]:
+    model_performances = {}
+
+    def evaluate_node(node_name, random_seed):
+        set_random_seed(random_seed)
+
+        node_data = data[node_name].to_numpy()
+        metric_evaluations = []
+        baseline_crps = {}
+        categorical = is_categorical(node_data)
+
+        if is_root_node(causal_model.graph, node_name):
+            for training_indices, test_indices in KFold(n_splits=kfolds, shuffle=True).split(node_data):
+                tmp_causal_mechanism = causal_model.causal_mechanism(node_name).clone()
+                tmp_causal_mechanism.fit(node_data[training_indices])
+
+                metric_evaluations.append(
+                    auto_estimate_kl_divergence(
+                        tmp_causal_mechanism.draw_samples(len(test_indices)), node_data[test_indices]
+                    )
+                )
+        else:
+            parent_data = data[get_ordered_predecessors(causal_model.graph, node_name)].to_numpy()
+
+            for training_indices, test_indices in KFold(n_splits=kfolds, shuffle=True).split(parent_data):
+                tmp_causal_mechanism = causal_model.causal_mechanism(node_name).clone()
+                tmp_causal_mechanism.fit(parent_data[training_indices], node_data[training_indices])
+
+                metric_evaluations.append(
+                    crps(parent_data[test_indices], node_data[test_indices], tmp_causal_mechanism.draw_samples)
+                )
+
+                if not compare_mechanism_baselines:
+                    continue
+
+                if not categorical:
```

Comment:
Nit (also no action required): Normally, it's more intuitive to start with `if categorical:`. Just one cognitive tick simpler ;-).